lightningd/invoice.c

```diff
@@ -125,6 +134,31 @@ static bool hsm_sign_b11(const u5 *u5bytes,
 	return true;
 }
 
+/* Return NULL if no error, or an error string otherwise. */
+static char *delete_invoice(const tal_t *cxt,
+			    struct wallet *wallet,
+			    struct invoices *invs,
+			    struct invoice *i)
+{
+	struct invoice_waiter *w;
+	if (!wallet_invoice_remove(wallet, i)) {
+		return tal_strdup(cxt, "Database error");
```

Comment:
Is there any reason you're allocating a string here vs just `return "Database error"` ?

Reply:
Flexibility, in case future errors require something like `tal_fmt` to express completely or in more detail.  Further the caller itself might use a long-lived context and need to explicitly `tal_free` the error. Current caller does not, admittedly. So to my mind, better to use the most flexible route.

I imagine, that in the future if expiration is implemented, it would have to use a longer-lived context, and it might call `delete_invoice` to delete expired invoices. Or maybe not.

Indeed it may be overdesigned, and in principle all that is needed is a bool flag return at the moment, as there is only one possible error for now.

Reply:
Yes, I would have just returned bool.  If there *were* multiple errors, a string is terrible if you have want to handle them differently, so we'd need to cross that bridge when we get to it anyway.  However, it's merely cosmetic here.